cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -0,0 +1,140 @@
+import co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTableAdmin;
+import co.cask.cdap.data2.transaction.queue.QueueAdmin;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.data2.util.hbase.TableId;
+import co.cask.cdap.proto.Id;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.client.HBaseAdmin;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Handles upgrade for System and User Datasets
+ */
+public class DatasetUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MDSUpgrader.class);
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final LocationFactory locationFactory;
+  private final QueueAdmin queueAdmin;
+  private final HBaseTableUtil hBaseTableUtil;
+  private final DatasetFramework namespacedFramework;
+
+  @Inject
+  public DatasetUpgrader(CConfiguration cConf, Configuration hConf, LocationFactory locationFactory,
+                         QueueAdmin queueAdmin, HBaseTableUtil hBaseTableUtil,
+                         @Named("namespacedDSFramework") DatasetFramework namespacedFramework) {
+
+    super(locationFactory);
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.locationFactory = locationFactory;
+    this.queueAdmin = queueAdmin;
+    this.hBaseTableUtil = hBaseTableUtil;
+    this.namespacedFramework = namespacedFramework;
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    // Upgrade system dataset
+    upgradeSystemDatasets(namespacedFramework);
+
+    // Upgrade all user hbase tables
+    upgradeUserTables();
+
+    // Upgrade all queue and stream tables.
+    queueAdmin.upgrade();
+  }
+
+  private void upgradeSystemDatasets(DatasetFramework framework) throws Exception {
+
+    // Upgrade all datasets in system namespace
+    Id.Namespace systemNamespace = Id.Namespace.from(Constants.SYSTEM_NAMESPACE);
+    for (DatasetSpecification spec : framework.getInstances(systemNamespace)) {
+      LOG.info("Upgrading dataset: {}, spec: {}", spec.getName(), spec.toString());
+      DatasetAdmin admin = framework.getAdmin(Id.DatasetInstance.from(systemNamespace, spec.getName()), null);
+      // we know admin is not null, since we are looping over existing datasets
+      admin.upgrade();
+      LOG.info("Upgraded dataset: {}", spec.getName());
+    }
+  }
+
+  private void upgradeUserTables() throws Exception {
+    DefaultDatasetNamespace namespace = new DefaultDatasetNamespace(cConf);
+    HBaseAdmin hAdmin = new HBaseAdmin(hConf);
+
+
+    for (HTableDescriptor desc : hAdmin.listTables()) {
+      String tableName = desc.getNameAsString();
+      TableId tableId = TableId.from(tableName);
```

Comment:
Can you add a TODO here that this will not work for upgrade from 2.8 to later versions? We will have to come up with a better way of mapping table name to dataset name and vice versa?